### ISSUE TITLE
Request to remove bias in pcap++ benchmarks

### ISSUE THREAD

**Issue by mfontanini** [seladb/PcapPlusPlus#562]
Hello! Co-author of [libtins](http://libtins.github.io/) here :wave:.

I've noticed that the benchmark results being displayed in the [pcap++ website](https://pcapplusplus.github.io/docs/benchmark#benchmark-1---tcp-parsing) are biased and I'd like to know your thoughts on it. For reference, all the benchmarks in that website are based on the [packet capture benchmarks repository](https://github.com/mfontanini/packet-capture-benchmarks) I created and that's the main reason why I'm writing here.

For all of the benchmarked libraries in my benchmarks I do the following:

1. Open a pcap file.
2. Read and parse every packet in that file.
3. Report total time taken

The idea is simple and is supposed to really measure one thing: how long do libraries take to parse all of those packets?

In my repository, I added a benchmark for [pcap++ itself](https://github.com/mfontanini/packet-capture-benchmarks/blob/0a766fdce60b5f408d71981af180b48dface6e4d/pcapplusplus/benchmark.cpp) a while ago. As the comment at the top states, I initially took this file from pcap++'s repository itself. You can see that exact same version of the file in this repo [here](https://github.com/seladb/PcapPlusPlus/blob/5d0d4a145f74c219275025b8e4affdeb038da402/Examples/PcapPlusPlus-benchmark/benchmark.cpp), as I literally copy/pasted it here.

It seems though that at some point in time though, the benchmark in the pcap++ repository was modified. The version now has a [subtle difference](https://github.com/seladb/PcapPlusPlus/blob/master/Examples/PcapPlusPlus-benchmark/benchmark.cpp#L86):

```diff
-                Packet packet(&rawPacket);
+                Packet packet(&rawPacket, pcpp::TCP);
```

This one line has big semantic implications on what the benchmark does: it now won't parse the entire packet, but it will instead stop after parsing the TCP layer. This means that, while all benchmarked libraries in my tests are parsing the entire packet into whatever class each library uses to represent a packet, this pcap++ benchmark is only parsing the ethernet, IP and TCP layers and ignoring the entire payload the packet contains after the TCP layer. For reference, the TCP packets generated in the benchmark contain a [742 byte payload](https://github.com/mfontanini/packet-capture-benchmarks/blob/master/benchmark.sh#L4) after the TCP layer, which accounts for over 90% of the size of the packet.

For a graphical representation, see this image:

![benchmarks](https://user-images.githubusercontent.com/969090/103957188-08d52100-50ff-11eb-9c20-7ad45dd466ce.png)

This means that now the benchmark in the pcap++ repository is doing less work than any other libaries in the benchmark framework I wrote. Now, if the benchmark involved the action "parsing a packet up to the TCP layer" then this would be a perfectly valid way of doing it and it's possible not many other libraries allow this kind of "limited parsing" of a packet, so pcap++ would likely be faster than any other of them. However, it is not: the point of this benchmark is to parse the entire packet, not a portion of it. If you think this is a common use case and want to add this as a separate benchmark, feel free to create a pull request into my benchmarks repository.

Now why I'm posting this: I've reached out through different means and never really got an answer on this. I think that, for a library that advertises itself as being very fast, it's a bit of a shame that a biased benchmark is the measure being used to back that statement up. Now I'm not saying the library isn't fast, it clearly is. In fact, a big round of optimizations I did on libtins was done after I first came across pcap++. It was in fact, at the time, faster than libtins and that drove me to improve the library and make a few optimizations that [under my benchmarks](http://libtins.github.io/benchmark/#bench1) made it faster than pcap++. Hell, it's possible that a newer version of pcap++ is faster than libtins, which I haven't touched in a while and therefore never re-ran the benchmarks. But that's not the point, the point is benchmarks should not be biased and should instead measure how fast the *exact same action* is performed by the different entities being evaluated.

What are your thoughts?

**Comment by seladb** [comment]
hi @mfontanini , thanks for reaching out!

I think we had this discussion in the past ([here](https://www.reddit.com/r/cpp/comments/95cpfy/pcapplusplus_v1808_released_a_multiplatform_c/)) and there were different opinions. I took a second look at the benchmark and I'm not sure how much difference there would be in parsing the payload of the packet. In the beginning I was worried that the generated payload may contain more protocols (hence eliminating the "apples-to-apples" comparison), but then I saw that the TCP ports are [always zero and the payload is always a series of 'A' characters](https://github.com/mfontanini/packet-capture-benchmarks/blob/0a766fdce60b5f408d71981af180b48dface6e4d/pcap_generator/generator.cpp#L31) so I think it should be fine.
I saw in [libtins benchmark page](http://libtins.github.io/benchmark/) that the latest version of PcapPlusPlus that was tested is 17.11 which is more than 3 years old already, may I suggest you run the benchmark again with a more recent version?

To be very honest, I'm not sure how this benchmark really matters to our users. We can probably agree at this point that the performance of both libraries is almost similar. I didn't see any PcapPlusPlus user complaining about the performance or claiming it's not good enough for their purposes. Did you encounter such a complaint from libtins users?
I built the `parseUntil` feature specifically for performance boost but I'm not sure how many users actually use it.

We can agree on how to perform the benchmark, but we may still get different results depending on the environment or other factors we can't control.
Maybe we can consider stop comparing the two libraries and focus more on metircs like the number of packets per second etc. (which is what the [DPDK is doing](https://core.dpdk.org/perf-reports/), for example). Please let me know what you think

**Comment by mfontanini** [comment]
> In the beginning I was worried that the generated payload may contain more protocols (hence eliminating the "apples-to-apples" comparison)

The fact that because of this the "parse up to layer" was introduced in the benchmark means that the benchmark became biased. In a normal scenario, users won't be parsing up to a layer, the default use case should be to parse a whole packet. By trying to avoid whatever the library's behavior could be (e.g. I presume you're trying to automatically parse the next layer given your comments here), now your benchmark started measuring a different use case than what the original benchmark aimed at.

> may I suggest you run the benchmark again with a more recent version

I definitely should do that. I haven't really payed much attention to libtins in a while and therefore never bothered to re-run them. As I said before, it's possible the benchmark ends up indicating pcap++ is faster but that's not why I created this issue: I'm not saying pcap++ isn't faster, I'm just saying you are measuring a different thing.

> We can probably agree at this point that the performance of both libraries is almost similar

Yeah, absolutely. The performance difference between the two is very small, regardless of which benchmark you're looking at. I don't imagine someone would choose one instead of the other because of what the benchmarks say. To your question, no, nobody has complained about the performance of it AFAIK.

> we may still get different results depending on the environment or other factors we can't control

Results will definitely be different depending on where we run them. The thing is, as long as all benchmarked libraries are actually doing the same, that's a fair comparison. e.g. while running these equivalent chunks of code in _this hardware_, this library performs X% slower/faster than this other one.

To be honest, now looking at what you're saying, I'm not sure what the implications of what the "parse until" change are and my graphic may be wrong. I just found it odd that a couple of weeks after libtins 4.0 (or whichever version included my performance improvements) was released, your benchmark was updated with what seemed like a change that could bias it and give it "an advantage" over the other libraries being compared against.

Anyway, I'll try to find some time and set up the environment to test all the libraries again and update the results on the benchmark page one of these days. Good to know your thoughts on this matter. Thanks!
